src/CodeGen_LLVM.cpp

```diff
@@ -1426,55 +1421,55 @@ void CodeGen_LLVM::visit(const Div *op) {
         return q - (rs & bs) + (rs & ~bs);
         */
 
-        Value *a = codegen(op->a), *b = codegen(op->b);
+        string a_name = unique_name('a');
+        string b_name = unique_name('b');
+        string q_name = unique_name('q');
+        string bs_name = unique_name("bs");
+        string rs_name = unique_name("rs");
+        Expr a = Variable::make(op->a.type(), a_name);
+        Expr b = Variable::make(op->b.type(), b_name);
+        Expr q = Variable::make(op->a.type(), q_name);
+        Expr r = a - q*b;
+        Expr bs = Variable::make(op->b.type(), bs_name);
+        Expr rs = Variable::make(r.type(), rs_name);
+        q = q - (rs & bs) + (rs & ~bs);
+        q = Let::make(rs_name, r >> (op->a.type().bits() - 1), q);
+        q = Let::make(bs_name, b >> (op->a.type().bits() - 1), q);
+        q = Let::make(q_name, div_round_to_zero(a, b), q);
+        q = Let::make(b_name, op->b, q);
+        q = Let::make(a_name, op->a, q);
```

Comment:
I hesitated to do that because if a and b are large expressions that have already run CSE, it would be running twice on most of that code. I also think that it's very unlikely for the lets to be unnecessary, because the only way that would happen is if a and b are the same expression (the rest of the expressions should be unique and used more than once).

That said, if you aren't concerned about the performance implications, I'd much rather do it that way, it's much less code and more readable.